_filters/execute_magick.py

```diff
@@ -0,0 +1,133 @@
+#!/bin/env python
+
+"""
+Pandoc filter to run all imagemagick code in code blocks
+"""
+
+import os
+import sys
+import hashlib
+from pandocfilters import toJSONFilter, Image, CodeBlock, Para
+from subprocess import Popen, PIPE
+
+DEBUG = False
+
+IMAGEDIR = '_images'
+
+# List of imagemagick tools; only commands in this list will be allowed to be run by the filter
+IM_COMMANDS = ('animate', 'compare', 'composite', 'conjure', 'convert', 'display', 'identify', 'import', 'mogrify', 'montage', 'stream', 'echo', 'printf')
+IM_IMAGE_TYPES = ('png', 'jpg', 'gif', 'tif')
+
+# Needed for writing unicode to stdout/stderr:
+reload(sys)
+sys.setdefaultencoding('utf-8')
+
+def sha1(x):
+    return hashlib.sha1(x.encode(sys.getfilesystemencoding())).hexdigest()
+
+def execute(code):
+    """
+    Executes the code.
+    Performs some sanity checks first.
+    """
+#    if DEBUG:
+#        sys.stderr.write('-'*100 + '\n')
+    code = remove_slash(code)
+    for line in code:
+        if DEBUG:
+            sys.stderr.write(line + u'\n')
+        commandline = line.split()
+        if commandline[0] not in IM_COMMANDS:
+            sys.stderr.write("Not a ImageMagick command: %s\n" % commandline[0])
+            return
+        image_name = commandline[-1]
+        image_type = os.path.splitext(image_name)[1][1:].lower()
+        if os.path.isfile(image_name):
+            sys.stderr.write("Image already exists: '%s'\n" % image_name)
+            return os.path.join('..', IMAGEDIR, image_name)
+        if image_type not in IM_IMAGE_TYPES:
+            sys.stderr.write("Not a ImageMagick Image Type: %s\n" % image_type)
+            return
+        p = Popen(' '.join(commandline), close_fds=True, stdout=PIPE, stderr=PIPE, shell=True)
+        out, err = p.communicate()
+        if out and DEBUG:
+            sys.stderr.write(out + '\n')
+        if err:
+            sys.stderr.write(err + '\n')
+    return os.path.join('..', IMAGEDIR, image_name)
+
+def remove_slash(text):
```

Comment:
Another YAGNI. If pandoc content is JSON, then we should use pythons internal `json.loads` to restore content. Codeblock content, as written, in the documents must be preserved.